### PR TITLE
🚀 [amp-story-player] Improve show() performance

### LINKED ISSUE #31212: [Story Player] Load player iframes sequentially when starting from other than first story
## Describe the new feature or change to an existing feature you'd like to see

The player loads too many stories at once and there is code to optimize the loading for the first story, but if you start with the second Story it takes forever to load.

We need to change the logic from:
`load the current story as visible + 2 others as prerender`
to:
`load the current story as visible, and once it's done, load the next one as prerender`

## Additional context

Publishers load a story using show()
- in that show() method, there's [this line](https://github.com/ampproject/amphtml/blob/master/src/amp-story-player/amp-story-player-impl.js#L755) that finds the adjascent stories 
- And [this line](https://github.com/ampproject/amphtml/blob/master/src/amp-story-player/amp-story-player-impl.js#L769) will start the logic that will position + preload the stories 
- the story will prerender, or load, when you call [this line](https://github.com/ampproject/amphtml/blob/master/src/amp-story-player/amp-story-player-impl.js#L1123) here 
- it's gated behind `navigationPromise` that's a `deferred` that waits for the visible story to send its `storyContentLoaded` event

Related to #31562, #31925

### PR DESCRIPTION
**Pull request by Enriqe**
To be merged after #31305
Closes #31212
Related to #31562, #31925

* This PR makes it so that when calling `show()` we load the current iframe before the next ones. 
* It also loads `MAX_IFRAMES_TO_LOAD` (which is 2 at the moment) instead of loading 3 iframes. It still positions the third iframe (just doesn't load any story on it).
* Adds unit testing for the iframe pool.

Change log:
1. Extract logic from `allocateIframeForStory_()` to update the iframe src and position. This makes it possible to position an iframe without necessarily loading anything on it. This is necessary for the new logic to work.
2. Renames `updateCurrentIframe_()` to  `layoutIframe_()`, and updates it so that it contains the logic from 1. This makes it reusable by codepaths that previously relied on the extracted logic from 1.
3. Changes logic from `IframePool.findAdjacent()` to use a BFS-like approach, where it returns an array that contains the current iframe first, then N+1, then N-1, then N+2 and so on... (Previously it would place the current one first, then N-2, N-1, N+1, N+2...).
3.1. The new `IframePool.findAdjacent()` also creates objects with more data about the iframe (the story idx it corresponds to, whether it should load, and its position). This is in preparation for moving more iframe logic from the `player-impl.js` to the `iframe-pool.js`. 
4. The new `loadStoriesInOrder_()` method that `show()` calls, uses the resulting queue from 3) and allocates, loads, and positions iframes accordingly.

### CODE REVIEW

File: src/amp-story-player/amp-story-player-iframe-pool.js [hunk 1]
```diff
@@ -14,6 +14,16 @@
...
 
+/** @enum {number} */
+const IframePosition = {
+  PREVIOUS: -1,
+  CURRENT: 0,
+  NEXT: 1,
+};
+
+/** @const {number} */
+const MAX_IFRAMES = 2;
```
Comment (gmajoulet) [inline src/amp-story-player/amp-story-player-iframe-pool.js:25]: You lost me here, why does the player have two `MAX_IFRAMES` constant that are different?

File: src/amp-story-player/amp-story-player-iframe-pool.js [hunk 2]
```diff
@@ -77,31 +87,59 @@ export class IframePool {
   }
 
   /**
-   * Finds adjacent iframe indices given an index.
-   * Examples of resulting adjacent arrays:
+   * Finds adjacent iframes given an story, starting in N and going to N+1.
+   * Following a DFS order.
```
Comment (gmajoulet) [inline src/amp-story-player/amp-story-player-iframe-pool.js:91]: What does this mean? An array is 2D so it can only be BFS?

File: src/amp-story-player/amp-story-player-iframe-pool.js [hunk 3]
```diff
@@ -77,31 +87,59 @@ export class IframePool {
...
-        ++cursor;
+      const cursorIdx = neighborsQueue.shift();
+
+      if (cursorIdx > maxIdx || cursorIdx < 0) {
+        continue;
+      }
+
+      adjacent.push({
+        storyIdx: cursorIdx,
+        shouldLoad: iframesToLoad > 0,
```
Comment (gmajoulet) [inline src/amp-story-player/amp-story-player-iframe-pool.js:122]: `iframesToLoad` quota will always be 0 once you get to `storyIdx - 1`, so you only prerender when navigating forward, is it a bug or a feature?

File: src/amp-story-player/amp-story-player-iframe-pool.js [hunk 4]
```diff
@@ -77,31 +87,59 @@ export class IframePool {
...
+   * positioned.
+   *
+   * Examples of traversal:
+   * idx
+   * 1 -> [2] [0] [1] [] []
    * 0 -> [0] [1] [2] [] []
-   * 1 -> [0] [1] [2] [] []
-   * 4 -> [] [] [0] [1] [2]
-   * 1 -> [0] [1]
+   * 4 -> [] [] [2] [1] [0]
```
Comment (gmajoulet) [inline src/amp-story-player/amp-story-player-iframe-pool.js:101]: for a MAX_IFRAMES of 2 these are wrong

File: src/amp-story-player/amp-story-player-iframe-pool.js [hunk 5]
```diff
@@ -77,31 +87,59 @@ export class IframePool {
...
+
+      if (cursorIdx > storyIdx && cursorIdx + 1 <= maxIdx) {
+        // Add neighbors to the right of the story to the queue.
+        neighborsQueue.push(cursorIdx + 1);
+        continue;
+      }
+
+      if (cursorIdx < storyIdx && cursorIdx - 1 >= 0) {
+        // Add neighbors to the left of the story to the queue.
+        neighborsQueue.push(cursorIdx - 1);
```
Comment (gmajoulet) [inline src/amp-story-player/amp-story-player-iframe-pool.js:140]: What do these lines do?

File: src/amp-story-player/amp-story-player-impl.js [hunk 6]
```diff
@@ -780,6 +781,48 @@ export class AmpStoryPlayer {
...
+      // First we must check whether the story has an iframe and allocate one
+      // if it doesn't.
+      if (!hasIframe) {
+        this.allocateIframeForStory_(currentIdx, navigatingBackwards);
+        iframeIdx = story.iframeIdx;
+      }
+
+      // Then we must load the current iframe and neighbor as visible and
+      // prerender respectively.
+      if (currentIframeObj.shouldLoad) {
```
Comment (gmajoulet) [inline src/amp-story-player/amp-story-player-impl.js:812]: With the current implementation, when navigating back you set the same `src` on the `NEXT` story

**Review by gmajoulet — COMMENTED** [review]
I'll be honest, this PR is too complex for me to understand or review properly. There is a lot I don't understand (cf comments) and the behavior is really hard to follow + not tested.
For example, is the fact that we don't prerender when navigating back a feature or a bug? How will we know if we break it?

Why does the iframe-pool decides of what has to be loaded? It should only rotate through iframe elements and decides of which ones to evict.

I think it's time to draw flowcharts and try to simplify the code paths to better support our use cases and performance strategy. The `priorityLoading` feature is good, but the iframe-pool is very confusing

File: src/amp-story-player/amp-story-player-impl.js (hunk 6 above)
Comment (Enriqe) [reply src/amp-story-player/amp-story-player-impl.js:812]: Sorry, not sure I understood. So you mean like calling `show()` and navigating to the previous story?

**Review by Enriqe — COMMENTED** [review]
> I'll be honest, this PR is too complex for me to understand or review properly. There is a lot I don't understand (cf comments) and the behavior is really hard to follow + not tested.

Yep my bad, will update with the PR with tests and flow charts. Hopefully will make it easier to understand.

> For example, is the fact that we don't prerender when navigating back a feature or a bug? How will we know if we break it?

Since we decided to only load 2 iframes at a time (current and prerender N+1), we don't prerender the previous one yeah. Unless there's no N+1, then we will prerender N-1.


> Why does the iframe-pool decides of what has to be loaded? It should only rotate through iframe elements and decides of which ones to evict.

(cf. description in PR) I think it would be great if we can move more iframe-specific logic to the iframe pool so not everything lives in `amp-story-player-impl.js`. This is a middle step in that direction.

**Comment by Enriqe** [comment]
Added unit tests for the `IframePool.findAdjacent()` (in addition to the behavioral tests in https://github.com/ampproject/amphtml/pull/31716) plus a flow diagram which should hopefully make it easier to understand.

Hope this makes it easier to review. PTAL @gmajoulet

**Comment by Enriqe** [comment]
Friendly ping @gmajoulet

File: src/amp-story-player/amp-story-player-iframe-pool.js [hunk 7]
```diff
@@ -77,31 +90,145 @@ export class IframePool {
...
+   *     |          +----------------------------------------+
+   *     |                               |
+   *     +-------------------------------+
+   *
    * @param {number} storyIdx
    * @param {number} maxIdx
-   * @return {!Array<number>}
+   * @return {!Array<!Object>}
    */
   findAdjacent(storyIdx, maxIdx) {
```
Comment (mszylkowski) [inline src/amp-story-player/amp-story-player-iframe-pool.js:181]: I think there's a simpler algorithm here to be used, since this is really complex:
```js
const searchCenter = storyIdx + 0.2; // Prefer next iframes more than previous, this number can be between 0 and 0.5
const indexDistancesToCenter = stories.map((story) => {story, distance: Math.abs(story.index - searchCenter)});
const closestStories = indexDistancesToCenter.sort((a, b) => a.distance - b.distance).slice(MAX_IFRAMES_TO_LOAD)
```

WDYT?

Comment (Enriqe) [reply src/amp-story-player/amp-story-player-iframe-pool.js:113]: Yeah doing a sort would degrade performance to O(nlog(n)) while we are doing O(n) currently.  That algo also doesn't distinguish between the iframes that need to be positioned or loaded.

I simplified it more by adding two cursors and moving them further away from the currentIdx. Is that easier to understand?

Comment (mszylkowski) [reply src/amp-story-player/amp-story-player-iframe-pool.js:113]: Much better 👍

**Review by gmajoulet — COMMENTED** [review]
Do we really need 3 constants, while loops, cursors, BFS traversing, queues, and logic split between two files to prerender the N+1 story?

I really struggle to understand why this code is so complicated, when we could simply say:
- on navigation to a story N, listen for story N load promise
- when story N is loaded, prerender story N+1
- if navigating to story X before story N loads, cancel the promise

If you think this PR is the way to go then I'll review it, lmk.

**Comment by Enriqe** [comment]
The current code is doing that. It loads story N, then prerenders N+1 after the load promise is done, or cancels if there's navigation (all of this using the existing `this.currentStoryLoadDeferred_`). 

I think you'd need to dig in a bit deeper to understand what the simple case you are suggesting is overlooking:

* We still need to position the 3rd iframe, even if we just load N and prerender N+1. So that if, after calling `show()` the user swipes to N-1, we load the corresponding story and run the corresponding animation when that happens.
* If e.g. there's no N+1 then we have to load N-1 and position N-2. If we write all these edge case conditionals in the `show()` method it would just make things harder to read and the code would grow even more.
* If we ever decide the number of iframes that need to be loaded or positioned, the code would be repetitive vs. just updating one constant.

Happy to go through the code in a meeting if that helps.

**Comment by gmajoulet** [comment]
> We still need to position the 3rd iframe, even if we just load N and prerender N+1. So that if, after calling show() the user swipes to N-1, we load the corresponding story and run the corresponding animation when that happens.

That code should be as simple as the snippet below. The fact that the codebase makes it MUCH harder than this is a big red flag, and we should take a step back and re-think the way this component is designed.

```
positionStories_() {
  [this.currentStoryIdx - 1, this.currentStoryIdx, this.currentStoryIdx + 1].forEach((idx, i) => {
    const story = this.stories_[idx];
    if (!story) return;
    story.iframe.setAttribute('distance', i - 1);
  })
}
```

> If e.g. there's no N+1 then we have to load N-1 and position N-2. If we write all these edge case conditionals in the show() method it would just make things harder to read and the code would grow even more.

I don't understand why you need to position N-2, it can't get into the view.

> If we ever decide the number of iframes that need to be loaded or positioned, the code would be repetitive vs. just updating one constant.

I don't agree with the idea of making our code very complicated and investing deeper in an architecture that, I think, we should move away from (this iframe pool is causing A LOT of complexity) for something we *might* do later.

I wrote a small design doc suggesting some design/architecture simplifications that I think we should tackle before looking at this PR again.

**Comment by Enriqe** [comment]
Closing in favor of #32153
